internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala

```diff
@@ -359,17 +352,18 @@ private[inc] abstract class IncrementalCommon(log: sbt.util.Logger, options: Inc
               inv("stamp changed from " + previousStamp + " to " + currentStamp)
           }
         def dependencyModified(file: File): Boolean =
-          previous.className(file) match {
-            case None => inv("no class name was mapped for it.")
-            case Some(name) => lookup.lookupOnClasspath(name) match {
-              case None    => inv("could not find class " + name + " on the classpath.")
-              case Some(e) => entryModified(name, e)
+          {
+            val classNames = previousRelations.libraryClassNames(file)
+            if (classNames.isEmpty) false
+            else classNames exists { binaryClassName =>
+              lookup.lookupAnalysis(binaryClassName) match {
```

Comment:
Yes, this will no longer detect shadowing by other libraries, which probably would have been caught before. One way to deal with it is to store the classpath and do the old `lookup.lookupOnClasspath`-based detection when the classpath changes.